iis/mymodule.cpp

```diff
@@ -693,146 +739,104 @@ CMyHttpModule::OnPostEndRequest(
     IN IHttpEventProvider * pProvider
 )
 {
-	REQUEST_STORED_CONTEXT *rsc = NULL;
-
-	rsc = (REQUEST_STORED_CONTEXT *)pHttpContext->GetModuleContextContainer()->GetModuleContext(g_pModuleContext);
+	RequestStoredContext* rsc = dynamic_cast<RequestStoredContext*>(pHttpContext->GetModuleContextContainer()->GetModuleContext(g_pModuleContext));
 
 	// only finish request if OnSendResponse have been called already
 	//
-	if(rsc != NULL && rsc->m_pResponseBuffer != NULL)
-	{
-		EnterCriticalSection(&m_csLock);
-
+	if(rsc != nullptr && rsc->ResponseBuffer() != nullptr)
+    {
+        CriticalSectionLock lock{cs};
```

Comment:
my lack of modern c++ probably... but where's the unlock?

Reply:
`CriticalSectionLock` is indeed just `std::unique_lock<T>` in disguise:
https://en.cppreference.com/w/cpp/thread/unique_lock

This class is specifically designed to acquire the lock in its constructor and release it, if held, upon its destruction. That's the cool part about idiomatic C++ RAII - whatever exit path the function takes, those wrappers make sure to release the resources properly because all the destructors will be called before the function returns to the caller.

In this particular case, the `CriticalSection` helper class added earlier implements BasicLockable requirements:
https://en.cppreference.com/w/cpp/named_req/BasicLockable

This makes sure it can be used along with `std::unique_lock`.

Another similar locking class is `std::lock_guard` but it is not that flexible and doesn't allow unlocking and then locking again without destructing the object.
Since we need to do it to run `modsecProcessRequest()` outside the scope of the lock, I opted for `std::unique_lock` instead.